### PR TITLE
Use custom hash table as node lookup table

### PR DESCRIPTION
**Pull request by rintaro**
Attempt to address the problem raised in #3 .

Use `Set<T>` like custom hash table that holds weak references for `RawSyntax` nodes.
This hash table doesn't holds `id`s separately, instead, use `T.id` via `Identifiable` protocol.
So freeing `RawSyntax` turns the bucket in the table into just `nil`, so it's reusable for referencing another object.

### CODE REVIEW

**Comment by rintaro** [comment]
@ahoppen @nkcsgexi 
What do you think?

**Comment by ahoppen** [comment]
I haven’t looked at the implementation in detail yet but the idea sounds great. Thanks for addressing this.

**Comment by nkcsgexi** [comment]
It's a great idea to use HashTable! My concern is why do we need to maintain our own implementation of it? can we define extensions on some existing stdlib or Foundation types to achieve the same goal?

**Comment by nkcsgexi** [comment]
I guess the custom part is "if the bucket contains nil, feel free to reuse it". I wonder if existing data structures can allow us to hack it.

**Comment by lorentey** [comment]
Yeah, the stdlib's hash table implementation does not currently support weak keys -- they need modified low-level lookup/insert/removal logic that the stdlib doesn't implement yet. It is not feasible to build a weak-keyed set on top of `Set`. (However, we should add a WeakSet and multiple WeakDictionary variants at some point!)

Dealloc'd entries need to work like tombstones -- lookups need to ignore them, while insertions can reuse their space, and removals/rehashes can compress them away. Implementation note: it's usually a bad idea to mutate storage on lookup operations -- people expect lookups to be thread-safe, and guaranteeing that could be difficult. (It would also mean that lookups could invalidate indices, which would be highly peculiar.)

In https://github.com/apple/swift/pull/19213, I'm working on a new `_HashTable` construct to unify low-level hash table operations across Set and Dictionary. (I expect to land that next week, after some additional work.) That struct would simplify creating a WeakSet/WeakDictionary within the standard library, but since it's internal to the stdlib, unfortunately it won't be directly helpful for SwiftSyntax. :(

**Comment by nkcsgexi** [comment]
Thank you for the explanation! @rintaro, @lorentey . If we have to include a non-trivial custom HashTable implementation in SwiftSyntax, i prefer we hold on landing this until we're sure the ever-increasing `nodeLookupTable` is an actual problem. With that being said, even if it is an actual problem, i prefer a simpler solution.

One potential solution is that server side periodically sends the syntax tree in full with a bit telling the clients to clear the `nodeLookupTable` all together. Thus we don't need to maintain the table forever in the SwiftSyntax side. What do you think? @ahoppen

**Comment by nkcsgexi** [comment]
I didn't mean to close it. Pushed the wrong button. Sorry!

**Comment by akyrtzi** [comment]
The hashtable implementation doesn't seem to be a huge amount of code and once the stdlib provides a WeakSet we'll just replace it with the stdlib's version and keep the same mode of operation.
If we introduce a different mode of operation, like periodically starting from scratch, and adding the heuristics to figure out when to do that, then this mode of operation should be replaced with the WeakSet model once the stdlib provides it.
Long-term it seems better to me to go with the weak-ref hashtable since it is a mode of operation that will be future-proof.

**Comment by lorentey** [comment]
I forgot to mention that Foundation already provides NSHashTable, which could be exactly what you need!

**Comment by rintaro** [comment]
> Implementation note: it's usually a bad idea to mutate storage on lookup operations -- people expect lookups to be thread-safe, and guaranteeing that could be difficult.

Makes sense. I removed mutation part from `subscript`.

> NSHashTable, which could be exactly what you need!

Thank you! I wasn't aware of `NSHashTable`. It seems `corelibs-foundation` doesn't have implementation of it though. So, perhaps what we should do is implementing `NSHashTable` in `corelibs-foundation`...

We will discuss about this next week.

**Comment by rintaro** [comment]
I will write unit tests for `WeakLookupTable` and merge this.

File: Sources/SwiftSyntax/SwiftSyntax.swift [hunk 1]
```diff
@@ -56,7 +59,14 @@ public final class SyntaxTreeDeserializer {
...
   /// Syntax nodes that have already been parsed and are able to be reused if
   /// they were omitted in an incremental syntax tree transfer
-  private var nodeLookupTable: [SyntaxNodeId: RawSyntax] = [:]
+  private var nodeLookupTable: WeakLookupTable<RawSyntax> = .init()
+
+  /// Keep a strong reference to the syntax tree that contains the nodes in the
+  /// `nodeLookupTable`. Because `nodeLookupTable` only holds a weak reference 
+  /// to the RawSyntax nodes, all retired `RawSyntax` nodes will be deallocated
+  /// once we set a new tree. The weak references in `nodeLookupTable` will then
+  /// become `nil` but will also never be accessed again.
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SwiftSyntax.swift]: I think this comment needs to be updated for `WeakLookupTable`

File: Sources/SwiftSyntax/WeakLookupTable.swift [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    return Int((Double(capacity) / maxLoadFactor).rounded(.up))
+  }
+
+  /// Number of bucket count to allocate to hold specified number of objects.
+  /// This is the next power of 2 greater than or equal to
+  /// '_minimalBucketCount(for: capacity)'
+  private static func _bucketCount(for capacity: Int,
+                                   from current: Int = 2) -> Int {
+    // Make sure it's representable.
+    precondition(capacity <= (Int.max >> 1) + 1)
```
Comment (ahoppen) [inline Sources/SwiftSyntax/WeakLookupTable.swift]: Why can't we have e.g. `capacity == Int.max - 1`?

File: Sources/SwiftSyntax/WeakLookupTable.swift [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    // Bucket count must always be power of 2.
+    precondition((current & (current - 1)) == 0)
+    // Minimum is 2 to guarantee at least 1 hole.
+    precondition(current >= 2)
+
+    let minimalBucketCount = _minimalBucketCount(for: capacity)
+    var bucketCount = current
+    while bucketCount < minimalBucketCount {
+      bucketCount &*= 2
```
Comment (ahoppen) [inline Sources/SwiftSyntax/WeakLookupTable.swift]: We don't expect any overflow here, right? Is the `&*=` just for performance reasons?

File: Sources/SwiftSyntax/WeakLookupTable.swift [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    }
+    return bucketCount
+  }
+
+  private var _bucketMask: Int {
+    @inline(__always) get {
+      return bucketCount &- 1
```
Comment (ahoppen) [inline Sources/SwiftSyntax/WeakLookupTable.swift]: Again is this `&-` just for performance reasons?

File: Sources/SwiftSyntax/WeakLookupTable.swift [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  }
+
+  @inline(__always)
+  private func _idealBucket(for id: Element.Identifier) -> Int {
+    return id.hashValue & _bucketMask
+  }
+
+  /// Finds the bucket where the object with the specified id should be stored
+  /// to.
+  private func _findHole(_ id: Element.Identifier) -> (pos: Int, found: Bool) {
```
Comment (ahoppen) [inline Sources/SwiftSyntax/WeakLookupTable.swift]: I think `alreadyExists` would be clearer in the return type than `found`. `found` implies that we found a hole which we will always do.

File: Sources/SwiftSyntax/WeakLookupTable.swift [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+    return true
+  }
+
+  /// Count the actual number of occupied buckets.
+  @inline(__always)
+  private func _countOccupiedBuckets() -> Int {
+    var count = 0
+    for i in 0 ..< bucketCount where buckets[i].value != nil {
+      count &+= 1
```
Comment (ahoppen) [inline Sources/SwiftSyntax/WeakLookupTable.swift]: Again `&+=` just for performance reasons? If so, I think we should document that. I always think why an overflow might make sense if I read a `&+` operator.

File: Sources/SwiftSyntax/WeakLookupTable.swift [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    }
+    return count
+  }
+
+  /// Reserves enough space to store a single new object. Returns true if
+  /// resizing happened.
+  private func _ensurePlusOneCapacity() -> Bool {
+    if bucketCount >= WeakLookupTable<Element>
+                        ._minimalBucketCount(for: estimatedCount &+ 1) {
```
Comment (ahoppen) [inline Sources/SwiftSyntax/WeakLookupTable.swift]: Same `&+`.

File: Sources/SwiftSyntax/WeakLookupTable.swift [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+      return false
+    }
+
+    // Slow path.
+    estimatedCount = _countOccupiedBuckets()
+    return reserveCapacity(estimatedCount &+ 1)
```
Comment (ahoppen) [inline Sources/SwiftSyntax/WeakLookupTable.swift]: `&+`

File: Sources/SwiftSyntax/WeakLookupTable.swift [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+    var (pos, found) = _findHole(obj.id)
+    if found {
+      return false
+    }
+
+    if /*resized=*/_ensurePlusOneCapacity() {
+      pos = _findHole(obj.id).pos
+    }
+    buckets[pos].value = obj
+    estimatedCount &+= 1
```
Comment (ahoppen) [inline Sources/SwiftSyntax/WeakLookupTable.swift]: `&+`

File: Sources/SwiftSyntax/WeakLookupTable.swift [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+  }
+
+  /// Get a object with specified id. Returns 'nil' if the object hasn't been
+  /// insert()-ed or it's already been freed.
+  public subscript(id: Element.Identifier) -> Element? {
+    // Since we don't fill the bucket when the object is freed (because we don't
+    // know), we can't stop iteration at a hole. So in the worst case (i.e. if
+    // the object doesn't exist in the table), full linear search is needed.
+    // However, since we assume the object exists and hasn't been freed yet,
+    // we expect it's stored near the 'idealBucket' anyway.
```
Comment (ahoppen) [inline Sources/SwiftSyntax/WeakLookupTable.swift]: We could think about adding a tombstone property to `WeakReference` so that we can detect when we found a hole. 
I.e.
WeakReference starts of as being `{occupied: false, value: nil}`. Once a value is set it becomes `{occupied: true, value: someValue}`. When `someValue` is freed it becomes `{occupied: true, value: nil}`. That way we can stop the search whenever we reach a `WeakReference` with `occupied == false` and thus avoid the full linear search.

However, in our use case, as you mentioned, we always expect to find the element, I think its worth it to make the trade-off to save an extra bit/byte for a linear worst-case complexity that we should never encounter in practice.

TL;DR: No need to change anything unless we are planning to merge this into `corelibs-foundation`.

File: Sources/SwiftSyntax/WeakLookupTable.swift [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+    let idealBucket = _idealBucket(for: id)
+    var bucket = idealBucket
+    repeat {
+      if let obj = buckets[bucket].value, obj.id == id {
+        return obj
+      }
+      bucket = (bucket &+ 1) & _bucketMask
```
Comment (ahoppen) [inline Sources/SwiftSyntax/WeakLookupTable.swift]: `&+`

File: Sources/SwiftSyntax/WeakLookupTable.swift (hunk 2 above)
Comment (rintaro) [reply Sources/SwiftSyntax/WeakLookupTable.swift]: Sorry, the condition was wrong.
The condition here is `_minimalBucketCount(for: capacity) <= (Int.max >> 1) + 1`.
As I added comments, the max bucket count here is `0b0100_0000_...` because `0b1000_0000_...` is out of bound.

File: Sources/SwiftSyntax/WeakLookupTable.swift (hunk 10 above)
Comment (rintaro) [reply Sources/SwiftSyntax/WeakLookupTable.swift]: Yeah, I don't intend to merge this into `corelibs-foundation`. :)
`NSHashTable` has very different API anyway.

**Comment by rintaro** [comment]
Thanks for detailed feedback @ahoppen !

> In particular I'd like to document whenever we are using &+ etc. just because of performance reasons because I usually expect to see a expected overflow whenever I see the unsafe operators. 

All `&+` family operations in this PR are only for performance. I added comments to justify the use of them.

> Also, do these actually make a performance difference?

Yes, it does affect the performance. I'm not sure how much, but using '&' operation is low-hanging-fruits to minimize the overflow check penalty.

File: Sources/SwiftSyntax/WeakLookupTable.swift [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+      count &+= 1
+    }
+    return count
+  }
+
+  /// Reserves enough space to store a single new object. Returns true if
+  /// resizing happened.
+  private func _ensurePlusOneCapacity() -> Bool {
+    // '&+' for performance. 'estimatedCount' is always less than 'bucketCount'
+    // which is 0x4000_... or below.
```
Comment (ahoppen) [inline Sources/SwiftSyntax/WeakLookupTable.swift]: This is not true. `estimatedCount` is always less than `buckecCount / maxLoadFactor` (it ends up giving the same guarantees for a `maxLoadFactor > 0.5` but still.

Consider the following example:

```
class Foo: Identifiable {
  let id: Int = 1
}

let lookupTable = WeakLookupTable<Foo>()
lookupTable.reserveCapacity(32)

while true {
  let newValue = Foo()
  lookupTable.insert(newValue)
  print(lookupTable.estimatedCount) /* need to make estimatedCount public for this test */
}
```

With each `insert` `estimatedCount` gets increased until `_minimalBucketCount(for: estimatedCount + 1)`, i.e. `(estimatedCount + 1) / maxLoadFactor` is greater than `bucketCount` at which point the slow path is taken and the lookup table realises that all the previous elements have been released and recounts.

This might influence some of the other comments as well. I haven't looked at them in detail.

Comment (rintaro) [reply Sources/SwiftSyntax/WeakLookupTable.swift]: > `estimatedCount` is always less than `buckecCount / maxLoadFactor` (it ends up giving the same guarantees for a `maxLoadFactor > 0.5` but still.

I believe this is "`estimatedCount` is alway less than or equal to `bucketCount * maxLoadFactor`". So, `'estimatedCount' is always less than 'bucketCount'` is true because `maxLoadFactor` is less than 1.

Either way, what we want to justify here is that `estimatedCount + 1` doesn't overflow. i.e. `estimatedCount < Int.max`, right?

Comment (ahoppen) [reply Sources/SwiftSyntax/WeakLookupTable.swift]: Ah, yes. I see. Sorry for the confusion. I somehow assumed that `bucketCount` is `32` in the example above, but its `64`. You're right and it definitely doesn't overflow.

**Comment by rintaro** [comment]
Squashed my commits.
@swift-ci Please test

**Comment by ahoppen** [comment]
Could you also close rdar://43516167 if you haven‘t done so already?
